### ISSUE TITLE
explore switching to pyright

### ISSUE THREAD

**Issue by Akuli** [Akuli/porcupine#560]
pyright seems to get bugs fixed pretty quickly, so it might be:
- better than mypy for type-checking
- better than pyls for default python langserver

**Comment by Akuli** [comment]
pyls is sometimes weird:

![Screenshot at 2021-07-09 17-40-42](https://user-images.githubusercontent.com/18505570/125095496-e8514800-e0dc-11eb-9b62-3154557a3de0.png)

Typeshed defines `getenv()` like this, and apparently pyls shows only one of the two overloads:

```python
@overload
def getenv(key: str) -> Optional[str]: ...
@overload
def getenv(key: str, default: _T) -> Union[str, _T]: ...
```

**Comment by Akuli** [comment]
More annoyances of pyls:
- Hover tooltips are huge. If a variable is a string, hovering it displays a description of what strings are, not something concise to indicate that the variable is a string. Pyright/pylance displays an awesome `foo: str` instead.
- I have contributed to typeshed a lot in the past couple years (especially to tkinter), but pyls ships with an ancient copy of typeshed. This means that I still see outdated hover tooltips.
- Autocompletions don't work when I annotate something as `list[str]` or `Foo | Bar`. These are newer typing syntax that didn't exist a few years ago.

tl;dr: pyls is ancient and was never that good to begin with.

**Comment by benjamin-kirkbride** [comment]
Would this be for integration within Porcupine, for porcupine dev (AKA what we use for linting, actions, etc), or both?

**Comment by rdbende** [comment]
Both

**Comment by Akuli** [comment]
Might be better to use basedpyright. It is a fork of pyright with more features that Microsoft intended to be closed-source and hence not part of pyright.

**Comment by benjamin-kirkbride** [comment]
See also:
- https://github.com/astral-sh/ty
- https://pyrefly.org/
